### DISCUSSION TITLE
Linker error with DM9051 driver

### DISCUSSION THREAD

**Discussion by technyon** [espressif/arduino-esp32#10139] [Q&A]
Hi,

I'm trying to add support for the DM9051 ethernet chip to my project. Everything compiles, but then I get a linker error (see below). I've enabled "CONFIG_ETH_SPI_ETHERNET_DM9051", the missing symbols in questions are esp_eth_mac_new_dm9051 and esp_eth_phy_new_dm9051. I'm not sure why the linker is missing these symbols, where are they supposed to come from?

The linker error only occurs when I call 

`ETH.begin(_type, _phy_addr, _cs, _irq, _rst, *_spi, _spi_freq_mhz);`

LAN8720  on the other hand works just fine.

The project is this, branch DM9051:

https://github.com/technyon/nuki_hub

```
/home/phr/.platformio/packages/toolchain-xtensa-esp32/bin/../lib/gcc/xtensa-esp32-elf/12.2.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/esp32dev/lib65e/libEthernet.a(ETH.cpp.o):(.literal._ZN8ETHClass8beginSPIE14eth_phy_type_tlPhiiiP8SPIClassiii17spi_host_device_th+0x30): undefined reference to `esp_eth_mac_new_dm9051'
/home/phr/.platformio/packages/toolchain-xtensa-esp32/bin/../lib/gcc/xtensa-esp32-elf/12.2.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/esp32dev/lib65e/libEthernet.a(ETH.cpp.o):(.literal._ZN8ETHClass8beginSPIE14eth_phy_type_tlPhiiiP8SPIClassiii17spi_host_device_th+0x34): undefined reference to `esp_eth_phy_new_dm9051'
/home/phr/.platformio/packages/toolchain-xtensa-esp32/bin/../lib/gcc/xtensa-esp32-elf/12.2.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/esp32dev/lib65e/libEthernet.a(ETH.cpp.o): in function `_ZN8ETHClass8beginSPIE14eth_phy_type_tlPhiiiP8SPIClassiii17spi_host_device_th':
/home/phr/.platformio/packages/framework-arduinoespressif32/libraries/Ethernet/src/ETH.cpp:641: undefined reference to `esp_eth_mac_new_dm9051'
/home/phr/.platformio/packages/toolchain-xtensa-esp32/bin/../lib/gcc/xtensa-esp32-elf/12.2.0/../../../../xtensa-esp32-elf/bin/ld: /home/phr/.platformio/packages/framework-arduinoespressif32/libraries/Ethernet/src/ETH.cpp:642: undefined reference to `esp_eth_phy_new_dm9051'
```

**Comment by Jason2866** [comment]
@technyon It is an Arduino as an IDF component project? I don't see enabling the DM9051 driver in sdkconfig.
BTW thanks for using my framework. Maybe a good idea to move to the Community pioarduino version.
This variant is 100% espressif Arduino compatible. You find it here https://github.com/pioarduino/platform-espressif32
There is an better (more correct) way to use the h2zero NimBLE lib. See https://github.com/pioarduino/platform-espressif32/tree/develop/examples/espidf-arduino-h2zero-BLE_scan

Have you made changes to `libsodum`? If not you can simply add via the espressif component manager. Same way as h2zero NimBLE library added in the BLE example.

REMOVE this lib https://github.com/technyon/nuki_hub/tree/master/lib/Ethernet It conflicts with the Arduino core 3.0.x!!

By using pioarduino you can remove all changes to get WifiClientSecure working. This is only needed since it is removed by design in my Tasmota framework. WifiClientSecure is there in the `pioarduino` build.

**Comment by Jason2866** [comment]
Why are you adding an entry for an IDF framework here? There is an actual version included in the platform.
```
[env:updater_esp32-c6]
extends = env:updater_esp32dev
board = esp32-c6-devkitm-1
platform_packages =
    platformio/framework-espidf @ https://github.com/tasmota/esp-idf/releases/download/v5.1.4.240629/esp-idf-v5.1.4.zip
```

**Comment by technyon** [comment]
Hi,

thanks for giving plenty of input :). Interestingly I've missed pioarduino so far, I'll have a look into it. 

Yes Arduino as an IDF component, you have to switch to the DM9051 branch, there "-DCONFIG_ETH_SPI_ETHERNET_DM9051" is added.

I think some things aren't setup perfectly yet because initially the project was using this cmake toolchain:

https://github.com/a9183756-gh/Arduino-CMake-Toolchain

I've just copied a lot of required libraries into my project, you're right they could now just be added as a dependency. The Ethernet library is still there to support the W5500 chip but, but as far as I've seen this should now be directly supported by ETH.h.

**Comment by Jason2866** [comment]
What happens if you add to sdkconfig additionaly this settings?
```
CONFIG_ETH_ENABLED=y
CONFIG_ETH_USE_SPI_ETHERNET=y 
```

**Comment by iranl** [comment]
It builds successfully with the following added in sdkconfig.defaults:

```
CONFIG_ETH_ENABLED=y
CONFIG_ETH_USE_SPI_ETHERNET=y
CONFIG_ETH_SPI_ETHERNET_DM9051=y
```

and removing the following from platformio.ini:

`-DCONFIG_ETH_SPI_ETHERNET_DM9051`

**Comment by technyon** [comment]
That indeed did the trick, thank you
